File: DSA11002 - CAY BIEU THUC 2.cpp
#include <bits/stdc++.h>
#define ll long long
#define boost ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
#define endl "\n"
#define MOD 1000000007

using namespace std;
int t = 1;

struct Node{
    string data;
    Node *left, *right;
};

Node *makeNode(string x){
    Node *newNode = new Node();
    newNode->data = x;
    newNode->left = newNode->right = NULL;
    return newNode;
}

Node *BuildTree(const vector<string> &s){
    int idx = 0, n = s.size();
    queue<Node*> q;
    Node *root = makeNode(s[idx++]);
    q.push(root);
    while(!q.empty() and idx < n){
        Node *tmp = q.front();
        q.pop();
        tmp->left = makeNode(s[idx++]);
        q.push(tmp->left);
        tmp->right = makeNode(s[idx++]);
        q.push(tmp->right);
    }
    return root;
}

ll calc(Node *root){
    if(!root->left and !root->right) return stoll(root->data);   // calc(root->left) and calc(root->right) can be replaced by left and right.
    if(root->data == "+") return calc(root->left) + calc(root->right);
    else if(root->data == "-") return calc(root->left) - calc(root->right);
    else if(root->data == "*") return calc(root->left) * calc(root->right);
    else return calc(root->left) / calc(root->right);
}


int main(){
    boost;
    cin >> t;
    while(t--){
        int n; cin >> n;
        vector<string> s(n);
        for(auto &x : s) cin >> x;
        Node *root = BuildTree(s);
        cout << calc(root) << endl;
    }
}

File: DSA11002 - CAY BIEU THUC 2.py
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def order(root):
    if not root.left and not root.right:
        return int(root.val)
    
    left_val = order(root.left)
    right_val = order(root.right)
    
    if root.val == '+':
        return left_val + right_val
    elif root.val == '-':
        return left_val - right_val
    elif root.val == '*':
        return left_val * right_val
    elif root.val == '/':
        return int(left_val / right_val)
        

t = int(input())
for _ in range(t):
    n = int(input())
    v = input().split()
    idx = 0
    root = Node(v[idx])
    idx += 1
    queue = [root]
    
    while queue and idx < n:
        node = queue.pop(0)
        if idx < n:
            node.left = Node(v[idx])
            queue.append(node.left)
            idx += 1
        if idx < n:
            node.right = Node(v[idx])
            queue.append(node.right)
            idx += 1
    
    print(order(root))
